### PR TITLE
mark the end of block also with STOP (v2)

### PR DESCRIPTION
**Pull request by hiqua**
Supersedes #52 

~Current status is that although it seems to solve the problem with toy contracts, there are new (different) failures with real-world contracts, so we should investigate further before merging these fixes. The diff is in e319c1a78fbfc11c05faf67b2103dcf287bfc77c.~

Should be ok now.

Depends on #70.

### CODE REVIEW

File: src/main/java/ch/securify/decompiler/DecompilerFallback.java [hunk 1]
```diff
@@ -100,9 +100,9 @@
 			for (int offset = 0; offset < rawInstructions.length; offset = ArrayUtil.nextNonNullIndex(offset, rawInstructions)) {
```
[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: I haven't look into it, I can do this later, but can you explain what the connection with this PR is?

File: src/main/java/ch/securify/decompiler/evm/OpCodes.java [hunk 2]
```diff
@@ -209,6 +209,30 @@ public static boolean isInvalid(int opcode) {
 		return "INVALID".equals(getOpName(opcode)) || opcode == INVALID;
 	}
 
+	/**
+	 * Indicate whether the opcode terminates a basic block
+	 * Ignores the case of JUMPI for which it depends on the CFG
```
[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/evm/OpCodes.java]: Please first see https://github.com/eth-sri/securify/pull/52#discussion_r237030935

To add to this, this PR just makes the current behavior more explicit with these new functions, it does not try to fix it (short of adding some missing opcodes), and that's intended.

File: src/main/java/ch/securify/decompiler/DecompilerFallback.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ritzdorf) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: In this PR you are fixing a piece of code that we don't need to fix, because we can simply remove it.

[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: OK I'll double-check that it's dead code.

[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: Well actually it's clear that it's not dead code, given that it does fix some bugs, e.g., with the empty contract and so on, and the other changes are simply introducing functions instead of listing the same opcodes every time. I'll still look into why this is needed.

[the remark this answers is not in the document]
Comment (ritzdorf) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: Thanks. Would be good to double-check and then we can remove it. 

Maybe introducing the fix you proposed for this for the Decompiler.java makes sense. (It has the same code.) Possibly fixing some function detection there. But that is part of a bigger open issue.

File: src/main/java/ch/securify/decompiler/evm/OpCodes.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ritzdorf) [reply src/main/java/ch/securify/decompiler/evm/OpCodes.java]: Ok. I agree that we don't need to reorganize everything in this PR, but why can't we add JUMPI to the instructions that ends a basic block, given that it does?

File: src/main/java/ch/securify/decompiler/DecompilerFallback.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: Yes, I've changed the code in Decompiler.java.

[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/DecompilerFallback.java]: I also double-checked, by running tests, and just by inspecting the code by removing useless instructions until I ended up with none, so it looks good to me.

File: src/main/java/ch/securify/decompiler/evm/OpCodes.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hiqua) [reply src/main/java/ch/securify/decompiler/evm/OpCodes.java]: I've removed this function instead, I'm waiting for @ptsankov's feedback about this weird condition in the Decompiler.
